src/appleseed/renderer/modeling/postprocessingstage/vignettepostprocessingstage.cpp

```diff
@@ -94,45 +100,49 @@ namespace
-            const Vector2f normalization_factor(lerp(resolution.y, resolution.x, m_anisotropy), resolution.y);
-
-            Image& image = frame.image();
-
-            for (std::size_t y = 0; y < props.m_canvas_height; ++y)
-            {
-                for (std::size_t x = 0; x < props.m_canvas_width; ++x)
-                {
-                    // Pixel coordinate normalized to be in the [-1, 1] range vertically.
-                    const Vector2f coord = (2.0f * Vector2f(static_cast<float>(x), static_cast<float>(y)) - resolution) / normalization_factor;
-
-                    //
-                    // Port of Keijiro Takahashi's natural vignetting effect for Unity.
-                    // Recreates natural illumination falloff, which is approximated by the "cosine fourth" law of illumination falloff.
-                    //
-                    // References:
-                    //
-                    //   https://github.com/keijiro/KinoVignette
-                    //   https://en.wikipedia.org/wiki/Vignetting#Natural_vignetting
-                    //
-
-                    const float linear_radial_falloff = norm(coord) * m_intensity;
-                    const float quadratic_radial_falloff = linear_radial_falloff * linear_radial_falloff + 1.0f;
-
-                    // Inversely proportional to the fourth power of the distance from the pixel to the image center.
-                    const float inverse_biquadratic_radial_falloff = 1.0f / (quadratic_radial_falloff * quadratic_radial_falloff);
-
-                    Color4f pixel;
-                    image.get_pixel(x, y, pixel);
-
-                    pixel.rgb() *= inverse_biquadratic_radial_falloff;
-
-                    image.set_pixel(x, y, pixel);
-                }
-            }
+
+            // Initialize effect-specific settings and context.
+            VignetteParams effect_params {
+                m_intensity,
+                m_anisotropy,
+                static_cast<float>(props.m_canvas_width),
+                static_cast<float>(props.m_canvas_height)
+            };
+
+            // Instantiate effect appliers, one per rendering thread.
+            VignetteApplierFactory effect_applier_factory(effect_params);
+            ImageEffectJob::EffectApplierVector effect_appliers;
+
+            effect_appliers.reserve(thread_count);
+            for (std::size_t i = 0; i < thread_count; ++i)
+                effect_appliers.push_back(effect_applier_factory.create());
+
+            // Create effect applier jobs.
+            ImageEffectJobFactory effect_job_factory;
+            ImageEffectJobFactory::EffectJobVector effect_jobs = effect_job_factory.create(
+                                                                    frame,
+                                                                    effect_appliers);
+
+            // Schedule effect applier jobs.
+            JobQueue job_queue;
+            for (const auto effect_job : effect_jobs)
```

Comment:
Is it a const reference ? (i.e.  `const auto&`).

also, we use const when the type can be easily determined like ` const auto something = static_cast<TYPE>(...)`. Maybe you should use explicit type here